### ISSUE TITLE
GPU enumeration intermittently fails when identical GPUs are present.

### ISSUE THREAD

**Issue by bb30994** [FoldingAtHome/fah-client-pub#1161]
Scenario: Linux, with two (identical) GTX 750 Ti's.  Client V7.4.9  Both slots are set to gpu "-1"

Sometimes the GPUs start up normally.  Sometimes one GPU starts up but the other one fails because it tries to use the index values which are already in use.  

Even if the slots are configured correctly the first time, this can change after a series of WUs are completed,  Because of the duplication, only one slot will be able to continue.

I've tried manually pre-setting index values but that, also, produces inconsistent results.  

Can there be a race condition during the configuration of the slots which can be resolved only by running mixed GPUs?  (I changed  one to a GTX 740 and have not had the problem since.)

**Comment by jcoffland** [comment]
It should work when you set two different GPU indices.  You should also see in the log file that the client is passing different ``-gpu #`` options to the cores.  Can you try this again and check the log files?

**Comment by bb30994** [comment]
...
> 04:29:26:******************************** Build ********************************
> 04:29:26:        Version: 7.4.9
> 04:29:26:           Date: Apr 27 2015
> 04:29:26:           Time: 06:59:43
> 04:29:26:     Repository: Git
> 04:29:26:       Revision: 0f91756f5844951be8a46310ba3d5b79256874b1
> 04:29:26:         Branch: master
> 04:29:26:       Compiler: GNU 4.8.4
> 04:29:26:        Options: -std=gnu++98 -O3 -funroll-loops -mfpmath=sse -ffast-math
> 04:29:26:                 -fno-unsafe-math-optimizations -msse2
> 04:29:26:       Platform: linux2 3.2.0-1-amd64
> 04:29:26:           Bits: 64
> 04:29:26:           Mode: Release
> 04:29:26:******************************* System ********************************
> 04:29:26:            CPU: Intel(R) Core(TM) i7 CPU 920 @ 2.67GHz
> 04:29:26:         CPU ID: GenuineIntel Family 6 Model 26 Stepping 4
> 04:29:26:           CPUs: 8
> 04:29:26:         Memory: 5.82GiB
> 04:29:26:    Free Memory: 4.62GiB
> 04:29:26:        Threads: POSIX_THREADS
> 04:29:26:     OS Version: 3.13
> 04:29:26:    Has Battery: false
> 04:29:26:     On Battery: false
> 04:29:26:     UTC Offset: -7
> 04:29:26:            PID: 2364
> 04:29:26:            CWD: /var/lib/fahclient
> 04:29:26:             OS: Linux 3.13.0-85-generic x86_64
> 04:29:26:        OS Arch: AMD64
> 04:29:26:           GPUs: 2
> 04:29:26:          GPU 0: Bus:3 Slot:0 NVIDIA:4 GM107 [GeForce GTX 750 Ti]
> 04:29:26:          GPU 1: Bus:3 Slot:0 NVIDIA:4 GM107 [GeForce GTX 750 Ti]
> 04:29:26:  CUDA Device 0: Platform:0 Device:0 Bus:3 Slot:0 Compute:5.0 Driver:7.5
> 04:29:26:  CUDA Device 1: Platform:0 Device:1 Bus:4 Slot:0 Compute:5.0 Driver:7.5
> 04:29:26:OpenCL Device 0: Platform:0 Device:0 Bus:3 Slot:0 Compute:1.2 Driver:352.63
> 04:29:26:OpenCL Device 1: Platform:0 Device:1 Bus:4 Slot:0 Compute:1.2 Driver:352.63
> 04:29:26:***********************************************************************
> 04:29:26:<config>
> 04:29:26:  
> 04:29:26:  <fold-anon v='true'/>
> 04:29:26:
> 04:29:26:  
> 04:29:26:  <gpu v='false'/>
> 04:29:26:
> 04:29:26:  
> 04:29:26:  <allow v='127.0.0.1,192.168.1.0/24'/>
> 04:29:26:
> 04:29:26:  
> 04:29:26:  <proxy v=':8080'/>
> 04:29:26:
> 04:29:26:  
> 04:29:26:  <command-allow-no-pass v='127.0.0.1,192.168.1.0/24'/>
> 04:29:26:  <password v='*****'/>
> 04:29:26:
> 04:29:26:  
> 04:29:26:  <power v='full'/>
> 04:29:26:
> 04:29:26:  
> 04:29:26:  <passkey v='********************************'/>
> 04:29:26:  <team v='131'/>
> 04:29:26:  <user v='borden.b'/>
> 04:29:26:
> 04:29:26:  
> 04:29:26:  <slot id='0' type='CPU'>
> 04:29:26:    <client-type v='beta'/>
> 04:29:26:    <cpus v='6'/>
> 04:29:26:    <max-packet-size v='big'/>
> 04:29:26:    <next-unit-percentage v='100'/>
> 04:29:26:  </slot>
> 04:29:26:  <slot id='1' type='GPU'>
> 04:29:26:    <next-unit-percentage v='100'/>
> 04:29:26:  </slot>
> 04:29:26:  <slot id='3' type='GPU'>
> 04:29:26:    <next-unit-percentage v='100'/>
> 04:29:26:  </slot>
> 04:29:26:</config>
> 04:29:26:Switching to user fahclient
> 04:29:26:Trying to access database...
> 04:29:26:Successfully acquired database lock
> 04:29:26:Enabled folding slot 00: READY cpu:6
> 04:29:26:Enabled folding slot 01: READY gpu:0:GM107 [GeForce GTX 750 Ti]
> 04:29:26:ERROR:FS03:'opencl-index'=0 is in use by another folding slot but
> GPU 1 matches this device's PCI bus= and PCI slot=
,,,
The way the error presents itself varies,(probably due to the results of some kind of 
timing/race-like differences during startup) but I think the stage was set during the initial setup.
...
> 04:55:13:<config>
> 04:55:13:  
> 04:55:13:  <fold-anon v='true'/>
> 04:55:13:
> 04:55:13:  
> 04:55:13:  <gpu v='false'/>
> 04:55:13:
> 04:55:13:  
> 04:55:13:  <allow v='127.0.0.1,192.168.1.0/24'/>
> 04:55:13:
> 04:55:13:  
> 04:55:13:  <proxy v=':8080'/>
> 04:55:13:
> 04:55:13:  
> 04:55:13:  <command-allow-no-pass v='127.0.0.1,192.168.1.0/24'/>
> 04:55:13:  <password v='*****'/>
> 04:55:13:
> 04:55:13:  
> 04:55:13:  <power v='full'/>
> 04:55:13:
> 04:55:13:  
> 04:55:13:  <passkey v='********************************'/>
> 04:55:13:  <team v='131'/>
> 04:55:13:  <user v='borden.b'/>
> 04:55:13:
> 04:55:13:  
> 04:55:13:  <slot id='0' type='CPU'>
> 04:55:13:    <client-type v='beta'/>
> 04:55:13:    <cpus v='6'/>
> 04:55:13:    <max-packet-size v='big'/>
> 04:55:13:    <next-unit-percentage v='100'/>
> 04:55:13:  </slot>
> 04:55:13:  <slot id='1' type='GPU'>
> 04:55:13:    <client-type v='beta'/>
> 04:55:13:    <max-packet-size v='40'/>
> 04:55:13:    <next-unit-percentage v='100'/>
> 04:55:13:  </slot>
> 04:55:13:  <slot id='2' type='GPU'>
> 04:55:13:    <client-type v='internal'/>
> 04:55:13:  </slot>
> 04:55:13:</config>
> 04:55:13:WU02:FS02:Connecting to 171.67.108.45:80
> 04:55:14:WU02:FS02:Assigned to work server 140.163.4.233
> 04:55:14:WU02:FS02:Requesting new work unit for slot 02: READY gpu:1:GM107
> [GeForce GTX 750 Ti] from 140.163.4.233
> 04:55:14:WU02:FS02:Connecting to 140.163.4.233:8080
> 04:55:14:WU02:FS02:Downloading 4.28MiB
> 04:55:20:WU02:FS02:Download 37.94%
> 04:55:26:WU02:FS02:Download 80.26%
> 04:55:28:WU02:FS02:Download complete
> 04:55:28:WU02:FS02:Received Unit: id:02 state:DOWNLOAD error:NO_ERROR
> project:10467 run:0 clone:273 gen:246 core:0x17 unit:0x000001ac538b3db9538bc096949cfc23
> 04:55:28:WU02:FS02:Starting
> 04:55:28:ERROR:WU02:FS02:Failed to start core: OpenCL device matching slot 2 not found
> 04:55:29:WU02:FS02:Starting
> 04:55:29:ERROR:WU02:FS02:Failed to start core: OpenCL device matching slot 2 not found
 ( now in a loop downloading WUs )
...

**Comment by bb30994** [comment]
There's another bug about client/control interactions when the log contains non-English characters.  Is this like causing that problem, too?
>>>> PCI bus=� and PCI slot=

**Comment by jcoffland** [comment]
Ok, I found and fixed the problem with printing the PCI bus and slot IDs incorrectly.  This will eliminate the non-English characters.

**Comment by jcoffland** [comment]
This part is definitely a problem:

```
> 04:29:26:          GPU 0: Bus:3 Slot:0 NVIDIA:4 GM107 [GeForce GTX 750 Ti]
> 04:29:26:          GPU 1: Bus:3 Slot:0 NVIDIA:4 GM107 [GeForce GTX 750 Ti]
```
The client has matched the same PCI bus and slot for both GPU folding slots whereas both the CUDA and OpenCL drivers show that there are two different GPUs:

```
> 04:29:26:  CUDA Device 0: Platform:0 Device:0 Bus:3 Slot:0 Compute:5.0 Driver:7.5
> 04:29:26:  CUDA Device 1: Platform:0 Device:1 Bus:4 Slot:0 Compute:5.0 Driver:7.5
> 04:29:26:OpenCL Device 0: Platform:0 Device:0 Bus:3 Slot:0 Compute:1.2 Driver:352.63
> 04:29:26:OpenCL Device 1: Platform:0 Device:1 Bus:4 Slot:0 Compute:1.2 Driver:352.63
```

**Comment by bb30994** [comment]
> This part is definitely a problem:

Is that part of NV's driver code or is it in your code?

Single-board/Dual-GPU configurations are widely tested but a pair of identical GPUs are much less common.  With a pair of single GPUs, they would be on separate bus IDs, unlike the single-board setup.

(I tried it with a 650-Ti and a 750-Ti but the 650 uses legacy drivers so they can't be combined on the same machine.  I do have a 740 that I can put with the 750TI and I bet it'll work.
... but the bug still needs to be fixed.

**Comment by bb30994** [comment]
I'm running 7.4.9.  If I manually set the index values to 0.0.0 and to 1,1,1 and pause/resume both GPUs they work.  Unfortunately the next time I restart FAHClient, those settings are removed from my config and then I often get the ERROR:'opencl-index'=0 is in use by another folding slot but GPU 1 matches this device's PCI bus=� and PCI slot= '  

IMHO we're looking at a race condition that assigns values for "opencl-index=-1" to two GPUs after the restart has successfully removed my "unnecessary" settings of 0,0,0 and 1,1,1 and changed everything back to -1,-1,-1.  

I'm willing to test your new FAHClient version if you've packaged it for Ubuntu.

**Comment by jcoffland** [comment]
Great work.  The new version is on the alpha download page.

**Comment by bb30994** [comment]
According to this information GPU enumeration order does not always match the enumeration order after a new reboot.   More information here: [https://foldingforum.org/viewtopic.php?f=38&t=29449](url)
but this is a more exhaustive test of 7.4.15 than I've seen elsewhere.

> ******************************* System ********************************
>CPU: Intel(R) Pentium(R) CPU G4400 @ 3.30GHz
>CPU ID: GenuineIntel Family 6 Model 94 Stepping 3
>CPUs: 2
>OS: Windows 10 Pro
>OS Arch: AMD64
>GPUs: 3
>GPU 0: Bus:4 Slot:0 NVIDIA:5 GP104 [GeForce GTX 1070]
>GPU 1: Bus:4 Slot:0 NVIDIA:5 GP104 [GeForce GTX 1070]
>GPU 2: Bus:4 Slot:0 NVIDIA:5 GP104 [GeForce GTX 1070]
>CUDA Device 0: Platform:0 Device:0 Bus:1 Slot:0 Compute:6.1 Driver:8.0
>CUDA Device 1: Platform:0 Device:1 Bus:2 Slot:0 Compute:6.1 Driver:8.0
>CUDA Device 2: Platform:0 Device:2 Bus:4 Slot:0 Compute:6.1 Driver:8.0
>OpenCL Device 0: Platform:0 Device:0 Bus:1 Slot:0 Compute:1.2 Driver:373.6
>OpenCL Device 1: Platform:0 Device:1 Bus:2 Slot:0 Compute:1.2 Driver:373.6
>OpenCL Device 2: Platform:0 Device:2 Bus:4 Slot:0 Compute:1.2 Driver:373.6
>OpenCL Device 3: Platform:1 Device:0 Bus:NA Slot:NA Compute:1.2 Driver:21.20
>OpenCL Device 4: Platform:1 Device:1 Bus:NA Slot:NA Compute:1.2 Driver:6.6,

**Comment by bb30994** [comment]
I DID NOT INTEND TO CLOSE THIS TICKET

**Comment by bb30994** [comment]
Now it's fixed in 7.4.16
